### PR TITLE
LINK-1169 create event external user

### PR DESCRIPTION
**Pull request by mikkojamG**
## Description :sparkles:

For external users / users in "other" organisation event form requires extra fields, and also some required fields that normally would be optional.

New fields:
- userEmail
- environmentalCertificate
- organization
- userPhoneNumber
- environment
- userConsent
- userName
- environment

Required fields for external user that are normally optional
- provider
- maximumAttendeeCapacity

Backend:
- https://github.com/City-of-Helsinki/linkedevents/pull/591

### Closes :no_good_woman:

**[LINK-1169](https://helsinkisolutionoffice.atlassian.net/browse/LINK-1169):**

## Additional notes :spiral_notepad:

[LINK-1169]: https://helsinkisolutionoffice.atlassian.net/browse/LINK-1169?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/domain/app/i18n/en.json [hunk 1]
```diff
@@ -736,16 +741,21 @@
...
-        "volunteering": "Volunteering organizer {{langText}} (if different from publisher)"
+        "volunteering": "Volunteering organizer {{langText}} (if different from publisher)",
+        "other": "Event organizer {{langText}}"
       },
       "labelPublisher": {
         "course": "Course publisher",
         "general": "Event publisher",
         "volunteering": "Volunteering publisher"
       },
+      "labelRegistrationLink": "Matkailun rekisteriselosteen URL",
```
Comment (laurakarhu) [inline src/domain/app/i18n/en.json:758]: pitäisi olla englanniksi kun on en.json tiedosto

File: src/domain/app/i18n/sv.json [hunk 2]
```diff
@@ -522,6 +522,7 @@
         "general": "Publicera evenemanget",
         "volunteering": "Publicera volontärarbetet"
       },
+      "buttonSendToPublishing": "Send to publising",
```
Comment (laurakarhu) [inline src/domain/app/i18n/sv.json:525]: ruotsiksi

File: src/domain/app/i18n/sv.json [hunk 3]
```diff
@@ -546,6 +547,7 @@
         "general": "Om antalet deltagare i evenemanget är begränsat, ange lägsta och högsta antal deltagare här.",
         "volunteering": "Om antalet deltagare i volontärarbetet är begränsat, ange lägsta och högsta antal deltagare här."
       },
+      "infoTextAttendeeCapacityUserUnknown": "If the number of event participants is not limited in any way, write a rough estimate of the maximum number of participants. (This information will be used for statistical purposes).",
```
Comment (laurakarhu) [inline src/domain/app/i18n/sv.json:550]: ruotsiksi

File: src/domain/app/i18n/sv.json
Comment (laurakarhu) [inline src/domain/app/i18n/sv.json:1]: Täällä sekä suomen että englanninkielisiä odottamassa käännöksiä ruotsiksi

File: src/utils/validationUtils.ts [hunk 4]
```diff
@@ -77,6 +77,9 @@ export const transformNumber = (
 export const isValidPhoneNumber = (phone: string): boolean =>
   /^\+?\(?\d{1,3}\)? ?-?\d{1,3} ?-?\d{3,5} ?-?\d{4}( ?-?\d{3})?/.test(phone);
 
+export const isValidEmail = (email: string): boolean =>
+  /^[A-Za-z0-9_!#$%&'*+/=?`{|}~^.-]+@[A-Za-z0-9.-]+$/i.test(email);
+
```
Comment (jorilindell) [inline src/utils/validationUtils.ts:82]: You could add unit tests to isValidEmail function

File: src/domain/event/validation.ts [hunk 5]
```diff
@@ -426,6 +434,92 @@ export const draftEventSchema = Yup.object().shape(
   CYCLIC_DEPENDENCIES
 );
 
+export const getExternalUserEventSchema = (
```
Comment (jorilindell) [inline src/domain/event/validation.ts:437]: There are quite many conditional validation rules. You could add unit tests to cover those cases.

File: src/domain/event/validation.ts [hunk 6]
```diff
@@ -426,6 +434,92 @@ export const draftEventSchema = Yup.object().shape(
...
+        .when([EVENT_FIELDS.PHONE_NUMBER], {
+          is: (phoneNumber: string) => !phoneNumber || phoneNumber.length === 0,
+          then: (schema) =>
+            schema.required(VALIDATION_MESSAGE_KEYS.STRING_REQUIRED),
+        })
+        .test(
+          'is-valid-email',
+          VALIDATION_MESSAGE_KEYS.EMAIL,
+          (value) => !value || isValidEmail(value)
+        ),
```
[the remark this answers is not in the document]
Comment (mikkojamG) [reply src/domain/event/validation.ts:497]: I was not aware that there is a build in method for emails 😁 This is much simpler, thanks 👍

[the remark this answers is not in the document]
Comment (mikkojamG) [reply src/domain/event/validation.ts]: Removed custom validation: [c75cd04](https://github.com/City-of-Helsinki/linkedcomponents-ui/pull/185/commits/c75cd044c10eb30000572d7391756cb80024568a)

File: src/domain/event/eventForm/EventForm.tsx [hunk 7]
```diff
@@ -472,20 +504,28 @@ const EventForm: React.FC<EventFormProps> = ({
...
             publisher: getValue(user?.organization, ''),
           },
-    [event, user]
+    [event, eventInitialValues, user?.organization]
   );
 
+  const validationSchema = !externalUser
+    ? publicEventSchema
+    : getExternalUserEventSchema();
+
```
[the remark this answers is not in the document]
Comment (mikkojamG) [reply src/domain/event/eventForm/EventForm.tsx:528]: Changed: [c75cd04](https://github.com/City-of-Helsinki/linkedcomponents-ui/pull/185/commits/c75cd044c10eb30000572d7391756cb80024568a)

File: src/domain/event/eventForm/EventForm.tsx [hunk 8]
```diff
@@ -472,20 +504,28 @@ const EventForm: React.FC<EventFormProps> = ({
 };
 
 const EventFormWrapper: React.FC<EventFormWrapperProps> = (props) => {
-  const { event } = props;
+  const { event, externalUser = false } = props;
   const { user } = useUser();
 
+  const eventInitialValues = !externalUser
+    ? EVENT_INITIAL_VALUES
+    : EVENT_UNKNOWN_USER_INITIAL_VALUES;
```
[the remark this answers is not in the document]
Comment (mikkojamG) [reply src/domain/event/eventForm/EventForm.tsx]: Changed: [c75cd04](https://github.com/City-of-Helsinki/linkedcomponents-ui/pull/185/commits/c75cd044c10eb30000572d7391756cb80024568a)

File: src/utils/validationUtils.ts (hunk 4 above)
Comment (mikkojamG) [reply src/utils/validationUtils.ts]: Removed custom validation: [c75cd04](https://github.com/City-of-Helsinki/linkedcomponents-ui/pull/185/commits/c75cd044c10eb30000572d7391756cb80024568a)

File: src/domain/event/validation.ts (hunk 5 above)
Comment (mikkojamG) [reply src/domain/event/validation.ts:436]: Added unit tests: [330dc3d](https://github.com/City-of-Helsinki/linkedcomponents-ui/pull/185/commits/330dc3dcde8d7a77badad592a10e94c7e3751d13)

File: src/domain/app/i18n/sv.json [hunk 9]
```diff
@@ -560,6 +561,7 @@
...
```
Comment (jorilindell) [inline src/domain/app/i18n/sv.json:564]: Should this be in Swedish?

File: src/domain/app/i18n/sv.json [hunk 10]
```diff
@@ -535,6 +535,7 @@
         "general": "Publicera evenemanget",
         "volunteering": "Publicera volontärarbetet"
       },
+      "buttonSendToPublishing": "Send to publising",
```
Comment (jorilindell) [inline src/domain/app/i18n/sv.json:538]: Should this be in Swedish?

File: src/domain/app/i18n/sv.json [hunk 11]
```diff
@@ -693,6 +695,10 @@
         "general": "Evenemangsbeskrivning {{langText}}",
         "volunteering": "Volontärarbetetsbeskrivning {{langText}}"
       },
+      "labelHasEnvironmentalCertificate": "Event has received an EcoCompass or similar environmental certification",
```
Comment (jorilindell) [inline src/domain/app/i18n/sv.json:698]: Should this be in Swedish?

File: src/domain/app/i18n/sv.json [hunk 12]
```diff
@@ -773,6 +782,8 @@
       "labelStartTime": "Starttid",
       "labelTime": "Tid",
       "labelUmbrellaEvent": "Huvudevenemang",
+      "labelUserConsent": "I have read <a href=\"https://www.hel.fi/static/liitteet-2019/Kaupunginkanslia/Rekisteriselosteet/Keha/Yritystoiminnan%20tukeminen%20ja%20neuvonta%20seka%20matkailuneuvontapalvelut.pdf\" target=\"_blank\" rel=\"noopener noreferrer\" aria-label=\"Matkailu register statement {{openInNewTab}}\">Matkailu register statement</a> and consent to use of my information",
```
Comment (jorilindell) [inline src/domain/app/i18n/sv.json:785]: Should this be in Swedish?

File: src/domain/event/formSections/responsibilitiesSection/ResponsibilitiesSection.tsx [hunk 13]
```diff
@@ -91,13 +95,18 @@ const ResponsibilitiesSection: React.FC<ResponsibilitiesSectionProps> = ({
         <FieldColumn>
           <MultiLanguageField
             disabled={!isEditingAllowed}
-            labelKey={`event.form.labelProvider.${type}`}
+            labelKey={
+              !isExternalUser
+                ? `event.form.labelProvider.${type}`
+                : 'event.form.labelProvider.other'
+            }
```
Comment (jorilindell) [inline src/domain/event/formSections/responsibilitiesSection/ResponsibilitiesSection.tsx:102]: I missed this in my first review round.  
```
labelKey={
              isExternalUser
                ? 'event.form.labelProvider.other'
                : `event.form.labelProvider.${type}`
            }
```
could be more readable

**Review by jorilindell — CHANGES_REQUESTED** [review]
It would be nice to have Swedish translations already at this point, event Swedish is not supported UI language. It's quite easy to forget to translate them later.

File: src/domain/event/formSections/responsibilitiesSection/ResponsibilitiesSection.tsx (hunk 13 above)
Comment (mikkojamG) [reply src/domain/event/formSections/responsibilitiesSection/ResponsibilitiesSection.tsx]: Fixed: [0b417dd](https://github.com/City-of-Helsinki/linkedcomponents-ui/pull/185/commits/0b417dd72d690a923be1b4e90dbbbbc74d68d4fa)

File: src/domain/app/i18n/sv.json (hunk 9 above)
Comment (mikkojamG) [reply src/domain/app/i18n/sv.json]: Translations: https://github.com/City-of-Helsinki/linkedcomponents-ui/pull/185/commits/9d8ee2af93331ff695811a6a1d684ae18200167c

File: src/domain/app/i18n/sv.json (hunk 10 above)
Comment (mikkojamG) [reply src/domain/app/i18n/sv.json]: Translations: https://github.com/City-of-Helsinki/linkedcomponents-ui/pull/185/commits/9d8ee2af93331ff695811a6a1d684ae18200167c

File: src/domain/app/i18n/sv.json (hunk 11 above)
Comment (mikkojamG) [reply src/domain/app/i18n/sv.json]: Translations: https://github.com/City-of-Helsinki/linkedcomponents-ui/pull/185/commits/9d8ee2af93331ff695811a6a1d684ae18200167c

File: src/domain/app/i18n/sv.json (hunk 12 above)
Comment (mikkojamG) [reply src/domain/app/i18n/sv.json]: Translations: https://github.com/City-of-Helsinki/linkedcomponents-ui/pull/185/commits/9d8ee2af93331ff695811a6a1d684ae18200167c

**Review by jorilindell — APPROVED** [review]
Good job! LGTM

**Pull request merged**